### ISSUE TITLE
First class Future object

### ISSUE THREAD

**Issue by jakirkham** [dask/dask#6380]
Currently with Distributed we have [`Future`s]( https://distributed.dask.org/en/latest/api.html#future ) in that context. However there doesn't exist a first class `Future` object within Dask itself. It would be useful to have a first class `Future` object within Dask. This would enable a wider set of algorithms and provided some needed flexibility for various problems people run into today.

First having a `Future` object would be useful in cases where we want to submit some work to start and then construct later operations with the knowledge that these results will soon become available (one example might be writing these to disk). Second this can be a useful way to prioritize elements of a workflow where we know what we want to prioritize (like working with large chunks of memory first). Third can help in cases where we want to interleave work between subtasks (recentering, normalizing, binning, etc.). Fourth this can be useful when the results of tasks are needed to more informatively generate new work (working with unknown chunks, selection based on data values, etc.). Fifth this can be a useful way of cutting down on large task graph handling issues (this is something other distributed libraries use https://github.com/dask/dask/issues/3514#issuecomment-390502785 ). There may be more ways this functionality could be used, but hopefully this provides some ideas on where/how it might help.

Additionally have gone through the issue tracker and identified issues/cases where this would either solve or help solve the underlying problem. This is not necessarily comprehensive. Also some issues may require more work on top of this. However in all cases this seems to provide a meaningful base to build on. Here's the list:

* https://github.com/dask/dask/issues/874
* https://github.com/dask/dask/issues/1519
* https://github.com/dask/dask/issues/2156 (at least practically)
* https://github.com/dask/dask/issues/3074 (useful for)
* https://github.com/dask/dask/issues/3255 (useful for)
* https://github.com/dask/dask/issues/3514
* https://github.com/dask/dask/issues/3587#issuecomment-396729211
* https://github.com/dask/dask/issues/3595#issuecomment-396971956
* https://github.com/dask/dask/issues/4156 (useful for)
* https://github.com/dask/dask/issues/4149 (useful for)
* https://github.com/dask/dask/issues/4544

Note this likely (though happy to be proven wrong) depends on moving to `concurrent.futures` as documented in issue ( https://github.com/dask/dask/issues/6220 ).

**Comment by mrocklin** [comment]
To be clear, are you suggesting implementing Futures like functionality for
the local scheduler?  I think that this would be a requirement.  I think
that it might require implementing an entirely new local scheduler.

No objections from me, but you might find that this is a moderately large
undertaking.

**Comment by jakirkham** [comment]
Yeah that's right. I agree. It does require some thought on what the local scheduler would look like. Am hoping this will be a bit easier with things like `concurrent.futures` ( https://github.com/dask/dask/issues/6220 ), but I could be wrong about that.

I understand. If you have suggestions on how we should approach this, would be open to them 🙂

**Comment by mrocklin** [comment]
My only suggestion is to make sure that it's worth your time before going
in.  Things will get complicated.
